src/driver/mod.rs

```diff
@@ -144,40 +144,75 @@ impl Drop for Driver {
-        let mut cancellable_ops = Vec::new();
-        for (id, cycle) in self.ops.lifecycle.iter() {
-            // don't cancel completed items
-            if !matches!(cycle, Lifecycle::Completed(_)) {
-                cancellable_ops.push(id);
+        // All Cancallable Ops are marked ignored
+        for (_, cycle) in self.ops.lifecycle.iter_mut() {
+            match std::mem::replace(cycle, Lifecycle::Ignored(Box::new(()))) {
+                lc @ Lifecycle::Completed(_) => {
+                    // don't cancel completed items
+                    *cycle = lc;
+                }
+
+                Lifecycle::CompletionList(indices) => {
+                    let mut list = indices.clone().into_list(&mut self.ops.completions);
+                    if !io_uring::cqueue::more(list.peek_end().unwrap().flags) {
+                        // This op is complete. Replace with a null Completed entry
+                        *cycle = Lifecycle::Completed(op::CqeResult {
+                            result: Ok(0),
+                            flags: 0,
+                        });
+                    }
+                }
+
+                _ => {
+                    // All other states need cancelling.
+                }
             }
         }
 
-        // cancel all ops
-        for id in cancellable_ops {
-            unsafe {
-                while self
-                    .uring
-                    .submission()
-                    .push(&AsyncCancel::new(id as u64).build().user_data(u64::MAX))
-                    .is_err()
-                {
-                    self.submit().expect("Internal error when dropping driver");
+        // Submit cancellation for all ops marked Ignored
+        for (id, cycle) in self.ops.lifecycle.iter_mut() {
+            if let Lifecycle::Ignored(..) = cycle {
+                unsafe {
+                    while self
+                        .uring
+                        .submission()
+                        .push(&AsyncCancel::new(id as u64).build().user_data(u64::MAX))
+                        .is_err()
+                    {
+                        self.uring
+                            .submit_and_wait(1)
+                            .expect("Internal error when dropping driver");
+                    }
                 }
             }
         }
 
-        // TODO: add a way to know if a multishot op is done sending completions
-        // SAFETY: this is currently unsound for multishot ops
-        while !self
-            .ops
-            .lifecycle
-            .iter()
-            .all(|(_, cycle)| matches!(cycle, Lifecycle::Completed(_)))
-        {
-            // If waiting fails, ignore the error. The wait will be attempted
-            // again on the next loop.
-            let _ = self.wait();
-            self.tick();
+        // Wait until all in flight cancellations have stopped
+        // Manually iterate the slab to process it in a single pass
+        let mut id = 0;
```

Comment:
That's a good idea. Manually iterate.

Reply:
I like it. Now my only concern would be the one I should have had for the earlier version too. ;) I'm going to accept but probably have a look later this afternoon at the lifecycle changes going on here. CompletionList to either Ignored or Completed, and here from Ignored to completed because tick is called, and anything else to being removed.

Nice.